plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/deploy/CloudSdkDeployProjectHandler.java

```diff
@@ -0,0 +1,171 @@
+                                   "Must select a primary runtime for " + project.getName());
+      return null;
+    }
+
+    boolean hasLoggedInUsers;
+    try {
+      hasLoggedInUsers = GCloudCommandDelegate.hasLoggedInUsers(project, runtime);
+    } catch (IOException | InterruptedException e) {
+      Activator.logAndDisplayError(null, TITLE, e.getMessage());
+      return null;
+    }
+
+    if (!hasLoggedInUsers) {
+      Activator.logAndDisplayError(null,
+                                   TITLE,
+                                   "Please sign in to gcloud before deploying project "
+                                          + project.getName());
+      return null;
+    }
+
+    final IPath warLocation = getWarLocationOrPrompt(project);
+    if (warLocation == null) {
+      Activator.logAndDisplayError(null,
+                                   TITLE,
+                                   "Must select the WAR directory to deploy " + project.getName());
+      return null;
+    }
+
+    final IPath sdkLocation = runtime.getLocation();
+    if (sdkLocation == null) {
+      Activator.logAndDisplayError(null, TITLE, "Set the location of " + runtime.getId());
+      return null;
+    }
+
+    Job job = new Job("Running Cloud SDK Deploy Action") {
+      @Override
+      protected IStatus run(IProgressMonitor monitor) {
+        try {
+          List<String> commands = new ArrayList<String>();
+          commands.add(sdkLocation + GCloudCommandDelegate.GCLOUD_DIR);
+
+          commands.add("preview");
+          commands.add("app");
+          commands.add("deploy");
+          commands.add(warLocation.toString());
+
+          MessageConsole messageConsole = MessageConsoleUtilities.getMessageConsole(project.getName()
+                                                                                    + " - "
+                                                                                    + TITLE,
+                                                                                    null);
+          messageConsole.activate();
+
+          IPath projectLocation = project.getLocation();
+
+          int exitCode = ProcessUtilities.launchProcessAndWaitFor(commands,
+                                                                  projectLocation.toFile(),
+                                                                  messageConsole.newMessageStream(),
+                                                                  null);
+
+          if (exitCode != 0) {
+            Activator.logAndDisplayError(null,
+                                         TITLE,
+                                         "Cloud SDK deploy action terminated with exit code "
+                                                + exitCode);
+          }
+        } catch (IOException | InterruptedException e) {
+          Activator.logError(e);
+        }
+        return Status.OK_STATUS;
+      }
+    };
+    job.schedule();
+    return null;
```

Comment:
@elharo it returns null from many places, not only from catch blocks and the last statement, so I think after combining the try blocks into a single block, a finally block that replaces only 2 lines, do so much for readability. I don't think it's worth the change. If we could find a better way to organize the method to avoid those repeated *1. get value, 2. give error if null* blocks, that would be better.